[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #31407 [quant][graphmode][refactor] Remove observers in the end
* **#31406 [quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`**

Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

jerryzh168 committed:
[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`

Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f9df4ad:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f9df4ad677c4d4773749a1f0b9b11a0fa4c115ce).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3997405) pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76340666) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76340666))
<details>
<summary>
<code>Dec 18 02:56:21 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 18 02:56:21 +++ eval 'extract_trap_cmd ' 
Dec 18 02:56:21 ++++ extract_trap_cmd 
Dec 18 02:56:21 ++++ printf '%s\n' '' 
Dec 18 02:56:21 +++ printf '%s\n' cleanup 
Dec 18 02:56:21 ++ trap -- ' 
Dec 18 02:56:21 cleanup' EXIT 
Dec 18 02:56:21 ++ which sccache 
Dec 18 02:56:21 ++ sccache --stop-server 
Dec 18 02:56:21 Stopping sccache server... 
Dec 18 02:56:21 error: couldn't connect to server 
Dec 18 02:56:21 caused by: Connection refused (os error 111) 
Dec 18 02:56:21 ++ true 
Dec 18 02:56:21 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 18 02:56:21 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Dec 18 02:56:21 ++ true 
Dec 18 02:56:21 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 18 02:56:21 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 18 02:56:21 ++ RUST_LOG=sccache::server=error 
Dec 18 02:56:21 ++ sccache --start-server 
Dec 18 02:56:21 Starting sccache server... 
Dec 18 02:56:22 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -637,7 +637,9 @@ class InsertQuantDeQuantHelper {
-  std::unordered_map<Graph*, std::unordered_map<Value*, QParamMap>> values_to_qparams_;
+  // Map from Graph to a map of output Value of forward call of observer to
+  // the original Value (input[1] of observer forward call)
+  std::unordered_map<Graph*, std::unordered_map<Value*, Value*>> values_to_quantize_;
```
Line seems a bit too long, do you mind running clang-format on it?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -637,7 +637,9 @@ class InsertQuantDeQuantHelper {
-  std::unordered_map<Graph*, std::unordered_map<Value*, QParamMap>> values_to_qparams_;
+  // Map from Graph to a map of output Value of forward call of observer to
+  // the original Value (input[1] of observer forward call)
+  std::unordered_map<Graph*, std::unordered_map<Value*, Value*>> values_to_quantize_;
```
I don't quite understand why we need this Value-Value map - the observer call node has all this information already. The only use the observer output has is as a key in the call to `getQSchemeAndQParamMap` - we should just use the original value as the key and then we won't need this ObserverOutput->OriginalValue map.

jerryzh168 replied:
`getQSchemeAndQParamMap` uses observer output to get the observer node actually, it's harder to get observer node from original value since in that case observer node is one of the uses of the original value, although it should be the only use.
Maybe I can store observer node here, and observer output would be n->output() and original value would be n->input(1)?

ZolotukhinM replied:
Yeah, that would be better.

ZolotukhinM reviewed (changes requested):
I think we can simplify this code a bit (it will probably require changes in the previous PR too): instead of having map of maps we should have a just a map. The inner map from observer-outputs to observer-original-value (or vice versa) is unnecessary. Also, once observers are removed some of the pointers become dead - they are only used as keys, so everything would still work, but it's generally not a great practice and in this particular case there is no need in this at all.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -637,8 +637,8 @@ class InsertQuantDeQuantHelper {
   std::unordered_map<Graph*, std::vector<Node*>> nodes_to_destroy_;
-  std::unordered_map<Graph*, std::unordered_map<Value*, QParamMap>>
-      values_to_qparams_;
+  // Map from Graph to observer node
```
That comment is not very helpful, it's basically repeats what's written on the line below. Please expand it a bit.

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`"


Summary:
Previously we record quantization parameters for a given value when we collect the observer nodes,
but actually the quantization parameter can vary depending on each module instance, to achieve
that, we need to delay the call to later stage and only record the `Value*` that's needed
in `collectObserverNodesAndValueToQuantize` function

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19162369](https://our.internmc.facebook.com/intern/diff/D19162369)

[ghstack-poisoned]
